### DISCUSSION TITLE
Importing Processing libraries

### DISCUSSION THREAD

**Discussion by tabreturn** [py5coding/py5generator#24] [Q&A] [answered]
How would I go about using Processing Java Libraries?

**Comment by hx2A** [comment, marked as answer]
Which Processing library would you like to use? Not all of them can be compatible, and not all of them are a good idea to use. For example, there are plenty of good Python physics libraries that can do what Box2D can do. Also, the Processing libraries will not be able to make arbitrary calls to your Python code. Libraries that need to do that won't work.

Here's an example of how you can use the Processing library Camera3D:

https://github.com/hx2A/py5examples/blob/main/examples/Using%20Camera3D.ipynb

The basic idea is to first put the Jars in a `jars` subdirectory (relative to Python's current working directory). If the Jar file is located somewhere else, use `py5_tools` to add your jar file's location to the classpath. `py5_tools` provides tools for getting the classpath or making changes to it. Note that when you import py5 it will also start the JVM. You need to make changes to the classpath before importing it.

Next you need to use `JClass` to make the Java class available to you in Python, as you see in the example. You can create an instance and pass it the Sketch object like this:

```
camera3D = Camera3D(py5.get_current_sketch())
```

Then you can continue on to use the Processing library to enhance your Sketch.

So far I know that PeasyCam, ColorBlindness, and Camera3D all work well with py5. (I created the latter 2, so I made sure they worked)

**Reply by tabreturn** [reply]
Great! Thanks, @hx2A

**Reply by hx2A** [reply]
@tabreturn I just finished the release. See the release notes for what's in it.

My brain is fried from writing/updating documentation all day. Still, I'm feeling pretty good about this release. Please test py5bot and the new settings/setup functionality and let me know how it works for you.

**Reply by tabreturn** [reply]
I had to install **libffi6** on my system (the latest Ubuntu-based Mint distro) using an old (deb) package from 2018. The 0.4 release was fine using **libffi7**. Anyhow, I've got both **libffi6** and **libffi7** on my system now, but I suspect that many Linux users won't be able to install libffi6 directly from their system repos.

**Reply by hx2A** [reply]
Can you tell me what kind of Python environment you are using? Anaconda? What error or problem arose that prompted you to install the older library?

I did fiddle with the package dependencies because some imports were missing. I didn't downgrade anything though. I'm not sure what the problem is.

**Reply by tabreturn** [reply]
Environment: Python 3.9.5 (which is built into Thonny 3.3.7). Error message --
```
%Run /home/nuc/Apps/thonny_py5_0.5/thonny/lib/python3.9/site-packages/py5_tools/tools/run_sketch.py py5.py
Traceback (most recent call last):
  File "/home/nuc/Apps/thonny_py5_0.5/thonny/lib/python3.9/site-packages/py5_tools/tools/run_sketch.py", line 22, in <module>
    from py5_tools import imported
  File "/home/nuc/Apps/thonny_py5_0.5/thonny/lib/python3.9/site-packages/py5_tools/__init__.py", line 23, in <module>
    from .imported import set_imported_mode, get_imported_mode  # noqa
  File "/home/nuc/Apps/thonny_py5_0.5/thonny/lib/python3.9/site-packages/py5_tools/imported.py", line 28, in <module>
    from .py5bot import py5bot
  File "/home/nuc/Apps/thonny_py5_0.5/thonny/lib/python3.9/site-packages/py5_tools/py5bot/__init__.py", line 20, in <module>
    from .kernel import Py5BotKernel  # noqa
  File "/home/nuc/Apps/thonny_py5_0.5/thonny/lib/python3.9/site-packages/py5_tools/py5bot/kernel.py", line 22, in <module>
    from ipykernel.zmqshell import ZMQInteractiveShell
  File "/home/nuc/Apps/thonny_py5_0.5/thonny/lib/python3.9/site-packages/ipykernel/__init__.py", line 2, in <module>
    from .connect import *
  File "/home/nuc/Apps/thonny_py5_0.5/thonny/lib/python3.9/site-packages/ipykernel/connect.py", line 12, in <module>
    import jupyter_client
  File "/home/nuc/Apps/thonny_py5_0.5/thonny/lib/python3.9/site-packages/jupyter_client/__init__.py", line 4, in <module>
    from .connect import *
  File "/home/nuc/Apps/thonny_py5_0.5/thonny/lib/python3.9/site-packages/jupyter_client/connect.py", line 21, in <module>
    import zmq
  File "/home/nuc/Apps/thonny_py5_0.5/thonny/lib/python3.9/site-packages/zmq/__init__.py", line 61, in <module>
    _load_libzmq()
  File "/home/nuc/Apps/thonny_py5_0.5/thonny/lib/python3.9/site-packages/zmq/__init__.py", line 23, in _load_libzmq
    import ctypes
  File "/home/nuc/Apps/thonny_py5_0.5/thonny/lib/python3.9/ctypes/__init__.py", line 8, in <module>
    from _ctypes import Union, Structure, Array
ImportError: libffi.so.6: cannot open shared object file: No such file or directory
```
I installed libffi6 (an older package for Ubuntu 2018.04) on my system and that solved the problem.

**Reply by hx2A** [reply]
Well, this is going to be a puzzle to solve.

Can you export your environment so I can see all of the package versions that are installed? ipykernel is one of the ones I bumped. Actually, I forgot to add it as a dependency before this weekend, but when I added it I first required >= 5.5. That caused problems on a Mac computer so I dropped it to >= 5.3 and then it seemed to be OK.

**Reply by tabreturn** [reply]
Cool, here you go -- a list for my v0.5 environment:
```
argon2-cffi         20.1.0
astroid             2.5.6
asttokens           2.0.5
async-generator     1.10
attrs               21.2.0
backcall            0.2.0
bcrypt              3.2.0
bitstring           3.1.7
bleach              3.3.1
certifi             2020.12.5
cffi                1.14.5
charset-normalizer  2.0.3
cryptography        3.4.7
debugpy             1.3.0
decorator           5.0.9
defusedxml          0.7.1
distro              1.5.0
docutils            0.17.1
ecdsa               0.16.1
entrypoints         0.3
esptool             3.0
idna                3.2
ipykernel           6.0.2
ipython             7.25.0
ipython-genutils    0.2.0
ipywidgets          7.6.3
isort               5.8.0
jedi                0.18.0
Jinja2              3.0.1
JPype1              1.3.0
jsonschema          3.2.0
jupyter-client      6.1.12
jupyter-core        4.7.1
jupyterlab-pygments 0.1.2
jupyterlab-widgets  1.0.0
lazy-object-proxy   1.6.0
line-profiler       3.3.0
MarkupSafe          2.0.1
matplotlib-inline   0.1.2
mccabe              0.6.1
mistune             0.8.4
mypy                0.812
mypy-extensions     0.4.3
nbclient            0.5.3
nbconvert           6.1.0
nbformat            5.1.3
nest-asyncio        1.5.1
noise               1.2.2
notebook            6.4.0
nptyping            1.4.2
numpy               1.21.1
packaging           21.0
pandas              1.3.0
pandocfilters       1.4.3
paramiko            2.7.2
parso               0.8.2
pexpect             4.8.0
pickleshare         0.7.5
Pillow              8.3.1
pip                 21.1.1
prometheus-client   0.11.0
prompt-toolkit      3.0.19
ptyprocess          0.7.0
py5                 0.5a0
pycparser           2.20
Pygments            2.9.0
pylint              2.8.2
PyNaCl              1.4.0
pyparsing           2.4.7
pyrsistent          0.18.0
pyserial            3.5
python-dateutil     2.8.2
pytz                2021.1
pyzmq               22.1.0
reedsolo            1.5.4
requests            2.26.0
Send2Trash          1.5.0
setuptools          56.0.0
six                 1.15.0
stackprinter        0.2.5
terminado           0.10.1
testpath            0.5.0
thonny              3.3.7
toml                0.10.2
tornado             6.1
traitlets           5.0.5
typed-ast           1.4.3
typing-extensions   3.10.0.0
typish              1.9.2
urllib3             1.26.6
wcwidth             0.2.5
webencodings        0.5.1
websockets          9.0.1
wheel               0.36.2
widgetsnbextension  3.5.1
wrapt               1.12.1
```
But then I noticed that my py5 v0.4 environment doesn't include the Jupyter packages in v0.5, which I suspect are looking for libffi.so.6 --
```
astroid            2.5.6
asttokens          2.0.5
backcall           0.2.0
bcrypt             3.2.0
bitstring          3.1.7
certifi            2020.12.5
cffi               1.14.5
charset-normalizer 2.0.3
cryptography       3.4.7
decorator          5.0.9
distro             1.5.0
docutils           0.17.1
ecdsa              0.16.1
esptool            3.0
idna               3.2
ipython            7.25.0
ipython-genutils   0.2.0
isort              5.8.0
jedi               0.18.0
JPype1             1.3.0
lazy-object-proxy  1.6.0
line-profiler      3.3.0
matplotlib-inline  0.1.2
mccabe             0.6.1
mypy               0.812
mypy-extensions    0.4.3
noise              1.2.2
nptyping           1.4.2
numpy              1.21.1
pandas             1.3.0
paramiko           2.7.2
parso              0.8.2
pexpect            4.8.0
pickleshare        0.7.5
Pillow             8.3.1
pip                21.1.1
prompt-toolkit     3.0.19
ptyprocess         0.7.0
py5                0.4a2
pycparser          2.20
Pygments           2.9.0
pylint             2.8.2
PyNaCl             1.4.0
pyserial           3.5
python-dateutil    2.8.2
pytz               2021.1
reedsolo           1.5.4
requests           2.26.0
Send2Trash         1.5.0
setuptools         56.0.0
six                1.15.0
stackprinter       0.2.5
thonny             3.3.7
toml               0.10.2
traitlets          5.0.5
typed-ast          1.4.3
typing-extensions  3.10.0.0
typish             1.9.2
urllib3            1.26.6
wcwidth            0.2.5
websockets         9.0.1
wheel              0.36.2
wrapt              1.12.1

```

**Reply by villares** [reply]
The easiest way in 2025 is now this! :smile:  

https://py5coding.org/how_tos/use_processing_libraries.html

**Comment by hx2A** [comment]
Here are the relevant differences:

```
jupyter-core        4.7.1
bleach              3.3.1
jupyterlab-pygments 0.1.2
widgetsnbextension  3.5.1
jsonschema          3.2.0
entrypoints         0.3
Jinja2              3.0.1
jupyterlab-widgets  1.0.0
pyparsing           2.4.7
terminado           0.10.1
pandocfilters       1.4.3
argon2-cffi         20.1.0
ipywidgets          7.6.3
nbformat            5.1.3
tornado             6.1
defusedxml          0.7.1
jupyter-client      6.1.12
nbclient            0.5.3
MarkupSafe          2.0.1
ipykernel           6.0.2
nbconvert           6.1.0
async-generator     1.10
packaging           21.0
mistune             0.8.4
notebook            6.4.0
pyrsistent          0.18.0
attrs               21.2.0
pyzmq               22.1.0
debugpy             1.3.0
testpath            0.5.0
nest-asyncio        1.5.1
prometheus-client   0.11.0
webencodings        0.5.1
```

Did you not try to use the py5 kernel in the previous release? I added the Jupyter related dependencies for the 0.5 release because the two Jupyter kernels both depend on them. The py5 kernel wouldn't have worked in your 0.4 setup.

I think perhaps that your 0.4 setup didn't care about libffi.so at all but the 0.5 setup does, and something was compiled against the older version of the library. I did some searching online and it seems this issue is something other libraries are dealing with also.

Also, you are using a virttual environment here, right? Not Anaconda? I believe an Anaconda environment would make it easy to install the right version of that shared library.

**Reply by hx2A** [reply]
I just checked my Anaconda environments and I see my py5 environments have libffi.so.7 in them but not libffi.so.6. Quite a few of the environments for other projects have both, however. So Anaconda seems to take care of this issue for users.

**Reply by tabreturn** [reply]
Yes, I'm not using Anaconda. I suspect this is something that'll get resolved over time, that won't affect all users, and we have a solution for now :+1:

**Reply by hx2A** [reply]
OK, cool.

Since I did a bad job on the release earlier this week I'd like to do another one fairly soon. I believe I can get the hex color thing done relatively quickly (and correctly), so I'll include it in the next release.

**Reply by hx2A** [reply]
And separately, I noticed you installed a bad version of jedi:

jedi                0.18.0

For that one the tab completion does not work correctly from Jupyter. I recommend installing 0.17.2 instead.

**Comment by jeremydouglass** [comment]
Re: using

`Camera3D = py5.JClass('camera3D.Camera3D')`

to import a Java class -> Python class -- have you considered using JImport instead? It appears to support the syntax `from <java class> import <Class> as <Classname>`

https://jpype.readthedocs.io/en/devel/imports.html

...and it looks like the behavior is [customizable](https://jpype.readthedocs.io/en/devel/api.html?highlight=importlib#jpype-imports-module) (although I haven't used JPype, I'm just looking through the docs.

**Reply by hx2A** [reply]
Hi @jeremydouglass 

Thanks for pointing out this feature. It does look like it can further simplify the bridge between Python and Java. It must be a new feature because it isn't available in 1.2.1, which is the version I have in front of me right now. The 1.3 version came out somewhat recently and I haven't upgraded my dev environment to use it yet.

Jpype is an awesome library with lots of great features, and I am doing my best to leverage them to add value to py5.

**Reply by hx2A** [reply]
@jeremydouglass this will be in the next release. I just had to add `import jpype.imports` to the py5 library, and now instead of writing this

```
Camera3D = py5.JClass('camera3D.Camera3D')
```

I can write this:

```
from camera3D import Camera3D
```

Much cleaner.

**Comment by villares** [comment]
I love PeasyCam, so I have to take the time to figure out the `p5_tools.add_jars()` thing, I'm following the [Camera3D example](https://github.com/hx2A/py5examples/blob/main/examples/Using%20Camera3D.ipynb), but no success yet...

**Reply by hx2A** [reply]
> I love PeasyCam, so I have to take the time to figure out the `p5_tools.add_jars()` thing, I'm following the [Camera3D example](https://github.com/hx2A/py5examples/blob/main/examples/Using%20Camera3D.ipynb), but no success yet...

I love PeasyCam also!

Are you using Thonny?

Try putting the PeasyCam jar in a jars subdirectory. That is easier than using `add_jars()`, but that is also a good option if the jars subdirectory is not working for you. Then use [get_classpath](http://py5.ixora.io/reference/get_classpath/) to see what the classpath is. It will show you the jars that it knows about. Also, note that if you are calling `add_jars()` you must do so before `import py5`. Since Thonny uses `run_sketch.py`, which does the import for you before any of your code gets executed, you can't use `add_jars()`. The jars subdirectory must be relative to the current working directory (`os.getcwd()`).

**Comment by villares** [comment]
This looks like you could be accessing an earlier version of py5 for some
reason? I remember you mentioned you have updated it  to 0.10.6... check if
there are multiple versions installed?
(On Thonny on Windows I saw it keep different versions one inside Thonny's
site-packages and another one at the users AppData if I remember correctly)

Em qua., 18 de jun. de 2025 às 21:49, vsquared ***@***.***>
escreveu:

**Reply by vsquared** [reply]
@villares For some reason updating py5 was more difficult on Windows11.  Thonny/manage packages would tell me that it had updated, but it didn't really do it.  I had to manually download the updated version, put it in AppData and then copy/paste py5 and py5_tools to replace the originals.  What a nightmare; **but** I really did get JavaFX installed.  Unfortunately now all I get is a non-functional FX2D window and the error message posted in my other thread.  We're getting closer to running py5 javafx in Windows 11 but not there yet.
